### PR TITLE
Search Document Generation

### PR DESCRIPTION
**Pull request by tomwilson-gs**
#### What type of PR is this?

Improvement

#### What does this PR do / why is it needed ?

Adds an ArtifactGenerationExtension and supplemental PURE code to generate documents for PackageableElements, ultimately destined for ingestion to ElasticSearch

#### Which issue(s) this PR fixes:

N/A

#### Does this PR introduce a user-facing change?

Yes, documents to be generated in build pipelines

### CODE REVIEW

File: legend-engine-xt-analytics-search-generation/src/main/java/org/finos/legend/engine/generation/SearchDocumentArtifactGenerationExtension.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import org.finos.legend.pure.generated.core_analytics_search_trans;
+import org.finos.legend.pure.generated.core_pure_protocol_protocol;
+import org.finos.legend.pure.m3.coreinstance.meta.pure.metamodel.PackageableElement;
+import org.slf4j.Logger;
+
+public class SearchDocumentArtifactGenerationExtension implements ArtifactGenerationExtension
+{
+    public final String ROOT_PATH = "searchDocuments";
+
+    private static final Logger LOGGER = org.slf4j.LoggerFactory.getLogger("Alloy Execution Server");
```
Comment (aormerod-gs) [inline legend-engine-xt-analytics-search-generation/src/main/java/org/finos/legend/engine/generation/SearchDocumentArtifactGenerationExtension.java:42]: Is this the right logger?

```suggestion
    private static final Logger LOGGER = org.slf4j.LoggerFactory.getLogger(SearchDocumentArtifactGenerationExtension.class);
```

File: legend-engine-xt-analytics-search-pure/src/main/resources/core_analytics_search/trans.pure [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+
+// CLASS DOCUMENT
+function meta::analytics::search::transformation::class::buildDocument(class: Class<Any>[1]): meta::analytics::search::metamodel::class::RootDocument[1]
+{
+  ^meta::analytics::search::metamodel::class::RootDocument(
+    id = $class->elementToPath(),
+    package = $class.package->toOne()->elementToPath(),
+    name = $class.name->toOne(),
+    type = 'alloyClass',
```
Comment (aormerod-gs) [inline legend-engine-xt-analytics-search-pure/src/main/resources/core_analytics_search/trans.pure:52]: this doesn't seem the correct type

File: legend-engine-xt-analytics-search-pure/src/main/resources/core_analytics_search/trans.pure [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+
+// SERVICE DOCUMENT
+function meta::analytics::search::transformation::service::buildDocument(service: meta::legend::service::metamodel::Service[1]): meta::analytics::search::metamodel::service::RootDocument[1]
+{
+  ^meta::analytics::search::metamodel::service::RootDocument(
+    id = $service.pattern,
+    name = $service.pattern,
+    type = 'alloyService',
```
Comment (aormerod-gs) [inline legend-engine-xt-analytics-search-pure/src/main/resources/core_analytics_search/trans.pure:167]: this doesn't seem the correct type

File: legend-engine-xt-analytics-search-pure/src/main/resources/core_analytics_search/metamodel.pure [hunk 4]
```diff
@@ -0,0 +1,104 @@
+Class meta::analytics::search::metamodel::BaseRootDocument
+[
+  idNotBlank: $this.id->length() > 0,
+  typeNotBlank: $this.type->length() > 0,
+  logicalTypeNotBlank: $this.logicalType->length() > 0
+]
+{
+  projectCoordinates : meta::analytics::search::metamodel::ProjectCoordinates[0..1];
+  id: String[1];
+  type: String[1];
```
Comment (aormerod-gs) [inline legend-engine-xt-analytics-search-pure/src/main/resources/core_analytics_search/metamodel.pure:10]: Should this be an enum, not a string

File: legend-engine-xt-analytics-search-pure/src/main/resources/core_analytics_search/metamodel.pure [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  logicalType: String[1];
```
Comment (aormerod-gs) [inline legend-engine-xt-analytics-search-pure/src/main/resources/core_analytics_search/metamodel.pure:11]: Does this belong in this model, or is it more of a UI concern?

Comment (tomwilson-gs) [reply legend-engine-xt-analytics-search-pure/src/main/resources/core_analytics_search/metamodel.pure]: Removed logicalType

**Review by aormerod-gs — APPROVED** [review]

**Pull request merged**